gh-91054: Add code object watchers API

itamaro opened the pull request:
This PR implements gh-91054, allowing extensions to set a callback function to be notified on creation and destruction of code objects

pyperformance results are 1.00x slower (~neutral) using AWS bare metal machine ([full comparison gist](https://gist.github.com/itamaro/fbcf5733d40fc3957d952f54d6d3e9e9))

Co-authored-by: @Ye11ow-Flash

itamaro committed:
gh-91054: Add API to allow extensions to set callback function on creation and destruction of PyCodeObject

Co-authored-by: Ye11ow-Flash <janshah@cs.stonybrook.edu>

itamaro committed:
Add tests for code object watcher

Co-authored-by: Ye11ow-Flash <janshah@cs.stonybrook.edu>

itamaro committed:
Add docs and news for code watchers

itamaro committed:
Merge branch 'main' into codeobject-watchers

itamaro committed:
Add Jaineel Shah (@Ye11ow-Flash) to ACKS list

markshannon commented on Objects/codeobject.c:
```diff
@@ -12,6 +12,62 @@
 
 
+static void
+notify_code_watchers(PyCodeEvent event, PyCodeObject *co)
```
Could you add a (8) bit vector of watchers, much as we did for function watchers.
That way, only one test is needed in the common case where there are no watchers.

itamaro replied:
sure, let me add it

itamaro committed:
Merge branch 'main' into codeobject-watchers

itamaro committed:
Add a bit vector to optimize watcher dispatch

A bit is set in the bit vector iff there is a watcher set at the
corresponding offset in the watcher array. Only notify watchers
if at least one bit is set.

markshannon commented on Objects/codeobject.c:
```diff
@@ -12,6 +12,66 @@
+notify_code_watchers(PyCodeEvent event, PyCodeObject *co)
+{
+    PyInterpreterState *interp = _PyInterpreterState_GET();
+    if (interp->active_code_watchers) {
```
I'm OK with this as is, but you could speed this up with an invariant.
`(active_code_watchers & (1 << i))` implies `interp->code_watchers[i] != NULL`

Then the loop can be something like this:
```
uint8_t watchers = interp->active_code_watchers;
while (watchers) {
    int watcher = left_most_bit(watchers);
    watchers &= ~(1 << watcher);
    if (interp->code_watchers[watcher](...) {
          PyErr_WriteUnraisable(...);
    }
}
```

This probably applies to the other watchers as well, so you might want to do this in another PR.

carljm replied:
I think this optimization applies to func watchers, but is already effectively applied for dict and type watchers; in those cases we already never look up the watcher unless the bit is set.

itamaro replied:
I have some more watchers-wide cleanup I wanted to do, so will add this optimization to the next PR, thanks!

itamaro replied:
cleanup and polish PR (including this optimization) is up at GH-99998

markshannon reviewed (approved):
Looks good. Does this complete the set of watchers?
